### PR TITLE
[WIP] Improved TypeScript support

### PR DESCRIPTION
**Pull request by ilyabo**

### CODE REVIEW

File: modules/dev-tools/templates/.eslintrc.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
+const typescriptConfigs = require('@typescript-eslint/eslint-plugin').configs;
+
+module.exports = {
+  parser: 'babel-eslint',
+  parserOptions: {
+    ecmaVersion: 2018
```
Comment (ibgreen) [inline modules/dev-tools/templates/.eslintrc.js:6]:
```suggestion
    ecmaVersion: 2021
```

File: modules/dev-tools/templates/.eslintrc.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  rules: {
+    'guard-for-in': 0,
+    'no-inline-comments': 0,
+    'camelcase': 0
+  },
+  overrides: [
+    {
+      files: ['**/*.ts', '**/*.tsx'],
+      parser: '@typescript-eslint/parser',
+      env: {browser: true, es6: true, node: true},
```
Comment (ibgreen) [inline modules/dev-tools/templates/.eslintrc.js:22]:
```suggestion
      env: {browser: true, es2021: true, node: true},
```
Why are we specifying the `envs` in this nested parserOptions and not up top?

File: modules/dev-tools/templates/.eslintrc.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    'prettier'
+  ],
+  rules: {
+    'guard-for-in': 0,
+    'no-inline-comments': 0,
+    'camelcase': 0
+  },
+  overrides: [
+    {
+      files: ['**/*.ts', '**/*.tsx'],
```
Comment (ibgreen) [inline modules/dev-tools/templates/.eslintrc.js:20]: Some comments on these overrides, why we do need these would be really helpful. You have done a lot of research here, you could help future maintainers by explaining e.g why we need `sourceType: module`

File: modules/dev-tools/templates/.eslintrc.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      parser: '@typescript-eslint/parser',
+      env: {browser: true, es6: true, node: true},
+      parserOptions: {
+        sourceType: 'module',
+        project: './tsconfig.json'
+      },
+      plugins: ['@typescript-eslint'],
+      rules: {
+        ...typescriptConfigs['eslint-recommended'].rules,
```
Comment (ibgreen) [inline modules/dev-tools/templates/.eslintrc.js:29]: Do you have any notes about which rules affect performance most. I still think performance is a bit of an issue with this plugin.

File: modules/dev-tools/templates/tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,23 @@
+{
+  "compilerOptions": {
+    "target": "es5",
```
Comment (ibgreen) [inline modules/dev-tools/templates/tsconfig.json:3]:
```suggestion
    "noEmit": "true",
```

File: modules/dev-tools/templates/tsconfig.json [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    "declaration": true,
+    "declarationMap": true,
+    "sourceMap": true,
+    "skipLibCheck": true,
+    "strict": true,
+    "noImplicitReturns": true,
+    "noFallthroughCasesInSwitch": true,
+    "forceConsistentCasingInFileNames": true,
+    "lib": [
+      "es2015",
```
Comment (ibgreen) [inline modules/dev-tools/templates/tsconfig.json:16]: Are the more recent available?

File: modules/dev-tools/package.json [hunk 7]
```diff
@@ -41,12 +41,16 @@
     "@babel/core": "^7.8.7",
     "@babel/plugin-transform-runtime": "^7.8.7",
     "@babel/preset-env": "^7.8.7",
+    "@typescript-eslint/eslint-plugin": "^4.14.1",
```
Comment (ibgreen) [inline modules/dev-tools/package.json:44]: We should bump the package.json version to `0.3.0`.

File: modules/dev-tools/templates/.eslintrc.js (hunk 3 above)
Comment (ibgreen) [inline modules/dev-tools/templates/.eslintrc.js:20]: I assume this also covers `d.ts` files?

**Review by ibgreen — APPROVED** [review]
Bump package.json to 0.3.0 to avoid automatic upgrades. 

Docs:
- Some kind of section about typescript support? 
- At least a bullet in what's new and upgrade-guide.md (I think renaming the eslint template my be a breaking change).
https://github.com/uber-web/ocular/tree/master/docs/dev-tools

File: modules/dev-tools/templates/.eslintrc.js (hunk 3 above)
Comment (ilyabo) [reply modules/dev-tools/templates/.eslintrc.js:22]: added a few

Comment (ilyabo) [reply modules/dev-tools/templates/.eslintrc.js:22]: It does

File: modules/dev-tools/templates/.eslintrc.js (hunk 4 above)
Comment (ilyabo) [reply modules/dev-tools/templates/.eslintrc.js:31]: Not really. I'd guess `recommended-requiring-type-checking` are probably the most expensive.

File: modules/dev-tools/templates/.eslintrc.js (hunk 2 above)
Comment (ilyabo) [reply modules/dev-tools/templates/.eslintrc.js:24]: Yes, good point. Probably this can be pulled up

File: modules/dev-tools/package.json (hunk 7 above)
Comment (ilyabo) [reply modules/dev-tools/package.json:44]: Done. Shouldn't we add `-alpha.0`?
BTW, can we test it without publishing.

**Comment by ibgreen** [comment]
Replaced by #361

**Pull request closed without merging**